hera_qm/vis_metrics.py

```diff
@@ -29,3 +30,173 @@ def check_noise_variance(data):
                     (data.channel_width * data.integration_time))
 
     return Cij
+
+
+def vis_bl_cov(uvd1, uvd2, bls, iterax=None):
+    """
+    Calculate visibility data covariance from uvd1 with uvd2
+    between specified baselines, optionally as a fuction of
+    frequency _or_ time.
+
+    Parameters
+    ----------
+    uvd1 : UVData object
+        A single-pol UVData object holding visibility data
+
+    uvd2 : UVData object
+        A single-pol UVData object holding visibility data
+
+    bls : list
+        A list of antenna-pair tuples or UVData baseline integers to
+        correlate.
+
+    iterax : str, optional
+        A data axis to iterate calculation over. Options=['freq', 'time'].
+
+    Returns
+    -------
+    blcov : ndarray
+        A covariance matrix of the data spanning Nbls x Nbls.
+        This ndarray is 4 dimensional, with shape (Nbls, Nbls, Ntimes^1, Nfreqs^2)
+            1 : Ntimes == 1 if iterax != 'time'
+            2 : Nfreqs == 1 if iterax != 'freq'
+    """
+    # type checks
+    assert isinstance(uvd1, UVData) and isinstance(uvd2, UVData), "uvd1 and uvd2 must be UVData objects"
+    assert uvd1.Npols == 1 and uvd2.Npols == 1, "uvd1 and uvd2 must be single-polarization objects"
+    assert isinstance(bls, (list, np.ndarray)), "bls must be a list of baselines"
+    if isinstance(bls[0], (int, np.integer)): bls = [uvd1.baseline_to_antnums(bl) for bl in bls]
+    assert iterax in [None, 'time', 'freq'], "iterax {} not recognized".format(iterax)
+    assert uvd1.Ntimes == uvd2.Ntimes, "Ntimes must agree between uvd1 and uvd2"
+    assert uvd1.Nfreqs == uvd2.Nfreqs, "Nfreqs must agree between uvd1 and uvd2"
+
+    # get Nfreqs and Ntimes
+    if iterax == 'time':
+        Ntimes = uvd1.Ntimes
+        Nfreqs = 1
+        sumaxes = (1,)
+    elif iterax == 'freq':
+        Ntimes = 1
+        Nfreqs = uvd1.Nfreqs
+        sumaxes = (0,)
+    elif iterax is None:
+        Ntimes = 1
+        Nfreqs = 1
+        sumaxes = (0, 1)
+
+    # construct empty cov
+    Nbls = len(bls)
+    cov = np.empty((Nbls, Nbls, Ntimes, Nfreqs), dtype=np.complex) * np.nan
+
+    # iterate over bls
+    for i, bl1 in enumerate(bls):
+        for j, bl2 in enumerate(bls):
+            # get data and weights
+            d1 = uvd1.get_data(bl1)
```

Comment:
one option to save on calling uvdata.get_data is by making four numpy arrays of shape (nbls, ntimes, nfreqs) (data and flags) for each uvd1 and uvd2. Compute the data mean and weights sums an multi-dimensional arrays, then compute the covariance by multiplying the arrays with newslices:

`cross_multiply = uvd1_data_array[None,...].conj() * uvd2_data_array[:, None, ...]` 
something like this creates (1,Nbls, Ntimes Nfreqs) and (Nbls, 1, Ntimes, Nfreqs) arrays which will be broadcast together to get the right shape. Would want to check the computation though, and the potential speed up. It would require some restructuring though and more thinking. maybe not worth it for the gain right now.